lib/rules/no-misleading-character-class.js

```diff
@@ -73,80 +72,120 @@ function isUnicodeCodePointEscape(char) {
-                (
-                    isUnicodeCodePointEscape(c1) ||
-                    isUnicodeCodePointEscape(c)
-                )
-            );
-        });
+    surrogatePair(char, index, chars) {
+        if (index === 0) {
+            return null;
+        }
+
+        const previous = chars[index - 1];
+
+        if (
+            isSurrogatePair(previous.value, char.value) &&
+            (
+                isUnicodeCodePointEscape(previous) ||
+                isUnicodeCodePointEscape(char)
+            )
+        ) {
+            return [previous, char];
+        }
+
+        return null;
     },
 
-    combiningClass(chars) {
-        return chars.some((c, i) => (
-            i !== 0 &&
-            isCombiningCharacter(c.value) &&
-            !isCombiningCharacter(chars[i - 1].value)
-        ));
+    combiningClass(char, index, chars) {
+        if (
+            index !== 0 &&
+            isCombiningCharacter(char.value) &&
+            !isCombiningCharacter(chars[index - 1].value)
+        ) {
+            return [chars[index - 1], char];
+        }
+
+        return null;
     },
 
-    emojiModifier(chars) {
-        return chars.some((c, i) => (
-            i !== 0 &&
-            isEmojiModifier(c.value) &&
-            !isEmojiModifier(chars[i - 1].value)
-        ));
+    emojiModifier(char, index, chars) {
+        if (
+            index !== 0 &&
+            isEmojiModifier(char.value) &&
+            !isEmojiModifier(chars[index - 1].value)
+        ) {
+            return [chars[index - 1], char];
+        }
+
+        return null;
     },
 
-    regionalIndicatorSymbol(chars) {
-        return chars.some((c, i) => (
-            i !== 0 &&
-            isRegionalIndicatorSymbol(c.value) &&
-            isRegionalIndicatorSymbol(chars[i - 1].value)
-        ));
+    regionalIndicatorSymbol(char, index, chars) {
+        if (
+            index !== 0 &&
+            isRegionalIndicatorSymbol(char.value) &&
+            isRegionalIndicatorSymbol(chars[index - 1].value)
+        ) {
+            return [chars[index - 1], char];
+        }
+
+        return null;
     },
 
-    zwj(chars) {
-        const lastIndex = chars.length - 1;
+    zwj(char, index, chars) {
+        if (
+            index !== 0 &&
+            index !== chars.length - 1 &&
+            char.value === 0x200d &&
+            chars[index - 1].value !== 0x200d &&
+            chars[index + 1].value !== 0x200d
+        ) {
+            return chars.slice(index - 1, index + 2);
+        }
 
-        return chars.some((c, i) => (
-            i !== 0 &&
-            i !== lastIndex &&
-            c.value === 0x200d &&
-            chars[i - 1].value !== 0x200d &&
-            chars[i + 1].value !== 0x200d
-        ));
+        return null;
     }
 };
 
-const kinds = Object.keys(hasCharacterSequence);
+const kinds = Object.keys(characterSequenceIndexFilters);
+
+/**
+ * Collects the indices where the filter returns true.
```

Comment:
The filters will return an array or `null`, so maybe like this?

```suggestion
 * Collects the indices where the filter returns an array.
```